### ISSUE TITLE
Dynamically created "input" with type checkbox does not work with ngModel 

### ISSUE THREAD

**Issue by Binvention** [angular/angular#7329]
in creating a options list that allows the user to change different settings i found a problem with the type checkbox . take this component 

```js
(function(app){
app.testing=ng.core.Component({
selector:"testing",
template:"<input type='checkbox' [(ngModel)]='testing'/>"+
"<template ngFor #item [ngForOf]='list'>"+
"<input [type]='item.type' [(ngModel)]='item.value'/>"+
"</template>"
}).Class({
constructor:[function(){
this.testing=true;
this.list=[
{type:"checkbox",value:true},
{type:"text",value:"hello world"}
];
}],
});
})(window.app||(window.app={}));
```

because of the ngFor loop combining with the dynamic setting of inputs and their types ngModel does not work on the second checkbox although it does on the first which is not dynamically set. im guessing this is a problem with how ngModel determines that the input is type checkbox. a example of this can be found [here](https://jsfiddle.net/Lakntoyv/4/) 

sorry to those of you who work in typescript i was not able to use typescript so it is in plain es5. i hope you can still interpret what its meaning and maybe someone can translate it to typescript.

**Comment by kemsky** [comment]
I don't know why this issue has such a low priority: you can not create extendable and reusable code, you are literally forced to do ctrl+c/v. You also can not extend components in angular, so the only choice is copy &paste pattern. It is really disappointing.

**Comment by zoechi** [comment]
Nobody said this issue has a low priority. There are just a lot of other issues with similar priority and there were a few huge tasks with higher priority.
You could wrap it in a reusable component with a matching `ControlValueAccessor`. Prefer composition over inheritance. Creating and composing components is the way to go in Angular2.

**Comment by ocombe** [comment]
You can find a plunkr reproducing this issue here: http://plnkr.co/edit/6X1itO2AO76bQmzzuX9Q?p=preview

**Comment by kemsky** [comment]
I guess official example does not work (it simply defaults to `text`), because it uses the same approach:
https://angular.io/docs/ts/latest/cookbook/dynamic-form.html

**Comment by pkozlowski-opensource** [comment]
Here is a plunker updated to RC.6: https://plnkr.co/edit/4FGmV1dmw5AwubEZNTDJ?p=preview

Now as much as I can see what you are trying to do this is kind of a corner case - you could easily get the desired behavior by using ngIf around 2 sets.

So yeh, while in the absolute it is a bug I don't think it is a blocker.

**Comment by kemsky** [comment]
Workaround is obvious only at first glance. Imagine common enterprise application with hundreds of inputs and forms, clearly you don't want to use low quality, error prone and verbose workarounds. Provided workaround is very limited in many ways: requires `ngSwitch` (bad code style and it is copy paste in the end), masking and formatting are impossible (`ngModel` feature), works good only with string fields (`ngModel` feature), changing type dynamically will change underlying DOM element (makes it difficult to integrate third party libraries, it might be not recommended, but implementing calendar or any other advanced control from scratch is not good idea either). It is impossible to create enterprise level controls using `ngModel` alone. At first i was so disappointed that decided to throw away forms module and implement it from scratch.

The only reasonable way i found includes wrapping inputs/selects into custom controls, implementing value accessors, internally avoid any `ngModel` stuff and carefully use events to watch for input changes (`Renderer` documentation is not very helpful by the way). It allows to decouple completely from input type, directives or any other artificially imposed constraints. You can implement formatting, input masks, bind 3rd party library and use any low level feature you want. However, custom controls can not be used efficiently in declarative way because validation can not be encapsulated into custom control, i.e. you have to specify the same email validator directive every time, but is already email(25 email fields all over the app): `<custom type="email" emailValidatorDirective>` (did not find way to overcome this). Other limitation of declarative way: `<custom type="email" [ngModel]="user?.email" (ngModelChange)="user == null || user.email = $event">` it confuses IDE, it is verbose but it is used in 100% of forms when you wait for object to load, for example C# is just fine with this: `doc?.setAttr('value')` and i believe it could be implemented for templates. So i ended up using `FormGroup` and custom form controls, which in turn requires additional mapping DTO <-> `FormGroup` absent out of the box, so i had to implement another form builder. 

So in my very-very humble opinion after 6 months with Angular, forms leave much to be desired despite of the fact that form is core component of the web-framework. 

Thank you for your attention.

**Comment by ewbutterfield** [comment]
I'm new to Angular2. I've been working on a creating a large form and creating dynamic input types was something that made a lot of sense for my usecase. While there are workarounds, I think this is something that could be run into a lot.

Here is a plunker that reproduces it in a simple way: https://plnkr.co/edit/O9tCGAbbwYWFYLpuRqKv?p=preview

**Comment by tom10271** [comment]
I am having exactly the same problem as @ewbutterfield mentioned. I don't want to mix template driven and model driven approach together. 

Really appreciated if it would be fixed.

**Comment by slavafomin** [comment]
We are developing new enterprise application and just stumbled upon this problem today (and we are just starting to use Angular). Do you have any estimates on when it's going to be fixed? Or could you at least provide some well-documented workarounds? Thank you!

**Comment by tytskyi** [comment]
@slavafomin if you want to have input with dynamic type, then you need to create wrapper component that will render inside input with proper type. This is a bit of manual work and trial and mistakes but in the end it worth it.

Then you will be able to do something like `<dynamic-type-input [type]="type" [(ngModel)]="value">`

There is no official docs at the moment, so you could look for example here:
https://blog.thoughtram.io/angular/2016/07/27/custom-form-controls-in-angular-2.html

**Comment by slavafomin** [comment]
Thank you @tytskyi for your suggestion. I've decided to just reference the `<input type="checkbox">` directly from my `CheckboxComponent` instead of referencing the generic `InputComponent` for now. This issue is not really a blocker, but an annoyance which takes time to figure out.

**Comment by thardy-pd** [comment]
I think this is still an issue.  It is a pain for us as well.

**Comment by artuska** [comment]
It is 2017 and Angular 4.4.4 and dynamic type for input fields does not work.

This works:

```
<input type="checkbox" formControlName="{{ parameter.key }}">
```

Those does not work:

```
<input type="{{ section.type }}" formControlName="{{ parameter.key }}">
or
<input attr.type="{{ section.type }}" formControlName="{{ parameter.key }}">
or
<input [type]="section.type" formControlName="{{ parameter.key }}">
```

Those 3 will create normal input field with `type="checkbox"` but binding will not work and `form.value` does not update when checkbox is clicked.

**Comment by keatkeat87** [comment]
<input [type]="'number'" formControlName="age" /> 
form.value.age will become String, not Number. 

<input type="number" formControlName="age" /> 
this work fine.

**Comment by TekTimmy** [comment]
It would be really nice to make it possible to have an dynamic type like `<input [type]="'number'" formControlName="age" />`

**Comment by funkizer** [comment]
<input [type]="'email'" [(ngModel)]="email" /> will also break email validation (both angular and browser native one) and validates as a text input. I wonder if it's because type is not a property of the element like most other attributes are and NG has created the element before property bindings / interpolation happens? :)

**Comment by trotyl** [comment]
@funkizer https://github.com/angular/angular/issues/15531#issuecomment-289555359

**Comment by kara** [comment]
This is a limitation of the framework itself - see https://github.com/angular/angular/issues/11716. We also have a forms-specific version of this issue here: https://github.com/angular/angular/issues/13243

**Comment by pauldraper** [comment]
> This is a limitation of the framework itself 

This is the place to file issues about the framework, yes?

Making a "reveal password" feature ought to take...maybe 2 lines?

```html
<input [type]="show.checked ? null : 'password'>
<input type="checkbox">
```

**Comment by trotyl** [comment]
@pauldraper Could you try it yourself first please?

https://stackblitz.com/edit/angular-cpep9d?file=app%2Fapp.component.html

What not being supported is dynamic `ControlValueAccessor` and your case doesn't require that at all.

**Comment by Maheshpanwar** [comment]
> 
> 
> in creating a options list that allows the user to change different settings i found a problem with the type checkbox . take this component
> 
> ```js
> (function(app){
> app.testing=ng.core.Component({
> selector:"testing",
> template:"<input type='checkbox' [(ngModel)]='testing'/>"+
> "<template ngFor #item [ngForOf]='list'>"+
> "<input [type]='item.type' [(ngModel)]='item.value'/>"+
> "</template>"
> }).Class({
> constructor:[function(){
> this.testing=true;
> this.list=[
> {type:"checkbox",value:true},
> {type:"text",value:"hello world"}
> ];
> }],
> });
> })(window.app||(window.app={}));
> ```
> 
> because of the ngFor loop combining with the dynamic setting of inputs and their types ngModel does not work on the second checkbox although it does on the first which is not dynamically set. im guessing this is a problem with how ngModel determines that the input is type checkbox. a example of this can be found [here](https://jsfiddle.net/Lakntoyv/4/)
> 
> sorry to those of you who work in typescript i was not able to use typescript so it is in plain es5. i hope you can still interpret what its meaning and maybe someone can translate it to typescript.

**Comment by Maheshpanwar** [comment]
https://github.com/angular/angular/issues/7329#issue-136877070

**Comment by MingyaoLiu** [comment]
Can confirm type Checkbox is still broken. 
When dynamically assign type checkbox the ngModel doesn't bind to the object. dynamically assign type number works tho.

live demo here: https://mingyaoliu.github.io/SVG-Polygon-Util/testPage

**Comment by bhushan90** [comment]
> I don't know why this issue has such a low priority: you can not create extendable and reusable code, you are literally forced to do ctrl+c/v. You also can not extend components in angular, so the only choice is copy &paste pattern. It is really disappointing.
> 
> Example:
> 
> ```
> <div><label [attr.for]="id">{{label}}</label></div>
> <div>
>     <input #input [attr.id]="id" [type]="type" [(ngModel)]="value" [attr.title]="tooltip" [readonly]="readonly" [class.datepicker]="type==='datetime'" />
> </div>
> ```
> 
> Currently i'm forced to COPY this template for each type of input(!) and to COPY corresponding component implementation (no 100% working way to inherit / no way to reuse with different template). It is just copy & paste in action. perhaps i could do some hacks, some dynamic loading, but should it be done this way?

Thank you so much I was facing the issue from a long time.

**Comment by MingyaoLiu** [comment]
> > I don't know why this issue has such a low priority: you can not create extendable and reusable code, you are literally forced to do ctrl+c/v. You also can not extend components in angular, so the only choice is copy &paste pattern. It is really disappointing.
> > Example:
> > ```
> > <div><label [attr.for]="id">{{label}}</label></div>
> > <div>
> >     <input #input [attr.id]="id" [type]="type" [(ngModel)]="value" [attr.title]="tooltip" [readonly]="readonly" [class.datepicker]="type==='datetime'" />
> > </div>
> > ```
> > 
> > 
> > Currently i'm forced to COPY this template for each type of input(!) and to COPY corresponding component implementation (no 100% working way to inherit / no way to reuse with different template). It is just copy & paste in action. perhaps i could do some hacks, some dynamic loading, but should it be done this way?
> 
> Thank you so much I was facing the issue from a long time.

If I remember correctly, the issue is not fixable in current angular design. Not sure if there are any new issue created for this, or if it's fixed in 8.0
